File: moderate/comb_sort.lua
local shrink = 1.25
for line in io.lines(arg[1]) do
  local a, n = {}, 0
  for i in line:gmatch("%d+") do
    a[#a+1] = tonumber(i)
  end
  local gap, last, changed = #a, 0, false
  while gap > 1 or changed do
    n, changed = n + 1, false
    if gap > 1 then
      gap = math.floor(gap / shrink)
    end
    for j = 1, #a - gap do
      if a[j] > a[j + gap] then
        a[j], a[j + gap], last, changed = a[j + gap], a[j], n, true
      end
    end
  end
  print(last)
end

File: moderate/comb_sort.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

static int shrink(int a) {
	return 4 * a / 5;
}

int main(int argc, char *argv[]) {
	FILE *fp;
	unsigned a, i = 0, ibs = 32, j, t;
	unsigned *ib = malloc(ibs * sizeof(unsigned));
	char c;

	if (argc != 2) {
		printf("Usage: %s [FILE]\n", argv[0]);
		return 1;
	}
	fp = fopen(*++argv, "r");
	while (fscanf(fp, "%u", &a) != EOF) {
		if (i == ibs) {
			ibs += ibs / 2;
			ib = realloc(ib, ibs * sizeof(unsigned));
		}
		ib[i++] = a;
		if ((c = getc(fp)) == '\n' || c == EOF) {
			unsigned n = 0, gap = i, last = 0;
			bool changed = false;
			while (gap > 1 || changed) {
				changed = false;
				n++;
				if (gap > 1)
					gap = shrink(gap);
				for (j = 0; j < i - gap; j++)
					if (ib[j] > ib[j + gap]) {
						t = ib[j];
						ib[j] = ib[j + gap];
						ib[j + gap] = t;
						changed = true;
						last = n;
					}
			}
			printf("%u\n", last);
			i = 0;
		}
	}
	free(ib);
	return 0;
}
